docs(recipes): add how to deploy modules to now.sh (#96)
americanexpress/one-app: docs/recipes/Deploying-Modules.md

## Deploying to Now with GitHub Actions

[Now.sh](https://zeit.co/home) allows you to deploy statics such as your modules. We will accomplish this be using [GitHub Actions](https://github.com/features/actions).

### Creating your Deploy Action

This deploy action uses Zeit's `now-cli` which you will need a [token from Now](https://zeit.co/account/tokens). The below action runs whenever a release is created or a push to master occurs. This will create 

```yml
name: Deploy

on:
  release:
    types: [created]
  push:
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js 12.x
      uses: actions/setup-node@v1
      with:
        node-version: '12.x'
    - name: Cache NPM Dependencies
      uses: actions/cache@v1
      with:
        path: ~/.npm
        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
        restore-keys: |
          ${{ runner.os }}-node-
    - name: Install Dependencies
      run: npm install
      env:
        NODE_ENV: development
    - name: Build Module
      run: npm run build
      env:
        NODE_ENV: production
    - name: 'Deploy'

>       run: now build --prod --confirm -t $NOW_TOKEN

Reviewer: Where is `now` coming from? Do I need to install `now-cli` to my package.json for this to work?

Author: Nope, now is installed in GitHub's virtual environments

const updateModuleMap = async () => {
  try {
    const response = await fetch(moduleMapUrl);

    const moduleMapContent = await response.json();
    const dir = 'tmp';

    moduleMapContent.modules[name] = {
      browser: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.browser.js`,
        integrity: bundleIntegrity.browser,
      },
      legacyBrowser: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.legacy.browser.js`,
        integrity: bundleIntegrity.legacyBrowser,
      },
      node: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.node.js`,
        integrity: bundleIntegrity.node,
      },
    };

    await fs.ensureDir(dir);
    await fs.writeFile(
      './tmp/module-map.json', JSON.stringify(moduleMapContent, null, 2)
    );
  } catch (error) {
    console.log(error);
  }
};

updateModuleMap().catch((err) => {
  // eslint-disable-next-line no-console
  console.log(err);
});
```

### Update your Deploy Action to Update your Module Map

At the bottom of your deploy action, you will want to add the following:

```yml
 - name: 'Update Module Map'
      run: |

>         node ./scripts/updateModuleMap.js

Reviewer: Maybe be explicit in where to put the `updateModuleMap.js` script inside `### Create an Update Module Script`

Reviewer 2: We need to have a broader discussion about where this script will live, it needs to be 100% reusable by all modules/static host providers and it should be either shipped with generator or added to the CLI

Reviewer: We will need to create tickets for evolving this Guide as this will become a multi-topic/multi-provider guide on running this. For now, starting small and simple is good but we will need to go places with this in the future.

Reviewer: Thanks for updating the description for `### Create an Update Module Script`

[👈 Return to Overview](./Recipes.md)

# Deploying Modules

## 📖 Table of Contents

* [Deploying to Now with GitHub Actions](#deploying-to-now-with-github-actions)
  * [Creating your Deploy Action](#creating-your-deploy-action)
* [Create and Update Your Module Map](#create-and-update-your-module-map)
  * [Creating a Module Map in Now.sh](#creating-a-module-map-in-nowsh)
  * [Create an Update Module Script](#create-an-update-module-script)
  * [Update your Deploy Action to Update your Module Map](#update-your-deploy-action-to-update-your-module-map)

## Deploying to Now with GitHub Actions

> [Now.sh](https://zeit.co/home) allows you to deploy statics such as your modules. We will accomplish this be using [GitHub Actions](https://github.com/features/actions).

Reviewer 3: ```suggestion
[Now.sh](https://zeit.co/home) allows you to deploy statics such as your modules. We will accomplish this by using [GitHub Actions](https://github.com/features/actions).
```

### Creating your Deploy Action

> This deploy action uses Zeit's `now-cli` which you will need a [token from Now](https://zeit.co/account/tokens). The below action runs whenever a release is created or a push to master occurs. The `now` command already comes available within GitHub actions, due to this you do not need to install this or add it to your `package.json`.

Reviewer 3: ```suggestion
This deploy action uses Zeit's `now-cli` so you will need a create [token with Now](https://zeit.co/account/tokens). The below action runs whenever a release is created or a push to master occurs. You do not need to install or add `now-cli` to your `package.json` as the `now` command is available within GitHub actions, .
```

      - master

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js 12.x
      uses: actions/setup-node@v1
      with:
        node-version: '12.x'
    - name: Cache NPM Dependencies
      uses: actions/cache@v1
      with:
        path: ~/.npm
        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
        restore-keys: |
          ${{ runner.os }}-node-
    - name: Install Dependencies
      run: npm install
      env:
        NODE_ENV: development
    - name: Build Module
      run: npm run build
      env:
        NODE_ENV: production
    - name: 'Deploy'
      run: now build --prod --confirm -t $NOW_TOKEN
      env:
        NOW_TOKEN: ${{ secrets.NOW_TOKEN }}
```

If you view the action running, during the `Deploy` step, you should see something like

```bash
Run now build --prod --confirm -t $NOW_TOKEN
Now CLI 17.1.1
- Setting up project
Linked to [YOUR NOW USERNAME]/[YOUR-REPO-NAME] (created .now and added it to .gitignore)
Inspect: https://zeit.co/[YOUR NOW USERNAME]/[YOUR-REPO-NAME]/
- Building
- Completing
Production: https://[YOUR-REPO-NAME].now.sh
https://[YOUR-REPO-NAME]-[HASH].now.sh
```

> ## Create and Update Your Module Map

Reviewer 2: I don't think we should expect people to create their own module map update script, it should be provided with two input variables 
1. MODULE_MAP_URL 
2. STATIC_ASSETS_URL

You will now need to add the module you just deployed to your module map.

### Creating a Module Map in Now.sh

> To create a module map in Now, you can create a module-map.json locally that looks something like this:

Reviewer 2: The same way we should provide the script to update the module map, we should provide one to create it for the first time, i.e. if it is a first time publish, I would suggest to generate it from the root module.

```json
{
  "modules": {

> 

Reviewer 4: Maybe a quick example or link to what the content looks like?
```
  [moduleName]: {
    node: { ... },
    browser: { ... },
    legacyBrowser: {}
  }
```

  }
}
```

You can deploy this by running the [now](https://zeit.co/docs/now-cli#commands/now/basic-usage) command within the directory that you created the `module-map.json` in. You can now view it at the production URL that `now` prints out, you will need that URL for the following steps.

> 

Reviewer 3: maybe also mention that they will need to update the module map url env var to point to where this is hosted

Author: Resolved in 7d6e942

```javascript
const fs = require('fs-extra');
const fetch = require('node-fetch');
const { name, version } = require('../package.json');

// This is created during the build process within the deploy action
const bundleIntegrity = require('../bundle.integrity.manifest.json');

const moduleMapUrl = 'YOUR MODULE MAP URL'; // This is the module map URL you got in the previous step

const STATIC_ASSETS_URL = 'URL TO WHERE YOUR MODULE IS DEPLOYED'; // example 'https://my-module.now.sh'

const updateModuleMap = async () => {
  try {
    const response = await fetch(moduleMapUrl);

    const moduleMapContent = await response.json();
    const dir = 'tmp';

    moduleMapContent.modules[name] = {
      browser: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.browser.js`,
        integrity: bundleIntegrity.browser,
      },
      legacyBrowser: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.legacy.browser.js`,
        integrity: bundleIntegrity.legacyBrowser,
      },
      node: {
        url: `${STATIC_ASSETS_URL}/${version}/${name}.node.js`,
        integrity: bundleIntegrity.node,
      },
    };

    await fs.ensureDir(dir);
    await fs.writeFile(

>       './tmp/module-map.json', JSON.stringify(moduleMapContent, null, 2)

Reviewer 2: not a fan of using the ./tmp folder to output the module map, we should find a more descriptive name for the folder (also the odd chance a CI also uses `tmp` and then you will end up with random artefacts alongside your module map)

[👈 Return to Overview](./Recipes.md)

# Deploying Modules

## 📖 Table of Contents

* [Deploying to Now with GitHub Actions](#deploying-to-now-with-github-actions)
  * [Creating your Deploy Action](#creating-your-deploy-action)
  * [Create and Update Your Module Map](#create-and-update-your-module-map)
    * [Creating a Module Map in Now.sh](#creating-a-module-map-in-nowsh)
    * [Create an Update Module Script](#create-an-update-module-script)
    * [Update your Deploy Action to Update your Module Map](#update-your-deploy-action-to-update-your-module-map)

> * [Deploying to Now with GitHub Actions](#deploying-to-now-with-github-actions)

Reviewer 5: Is this just a dupe of line 7?

Author: Thanks for catching that... My TOC creator in VSCode must of not updated correctly and I missed that.

> # Deploying to Now with GitHub Actions

Reviewer 5: this is the wrong heading level, should be h2 - adjust all below to match

[Now.sh](https://zeit.co/home) allows you to deploy statics such as your modules. We will accomplish this be using [GitHub Actions](https://github.com/features/actions).

## Creating your Deploy Action

This deploy action uses Zeit's `now-cli` which you will need a [token from Now](https://zeit.co/account/tokens). The below action runs whenever a release is created or a push to master occurs. This will create 

```yml
name: Deploy

on:
  release:
    types: [created]
  push:
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js 12.x
      uses: actions/setup-node@v1
      with:
        node-version: '12.x'
    - name: Cache NPM Dependencies
      uses: actions/cache@v1
      with:
        path: ~/.npm
        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
        restore-keys: |
          ${{ runner.os }}-node-
    - name: Install Dependencies
      run: npm install
      env:
        NODE_ENV: development

>     - name: Build One App Page Wrapper

Reviewer 5: What's "One App Page Wrapper"?

      run: npm run build
      env:
        NODE_ENV: production
    - name: 'Deploy'
      run: now build --prod --confirm -t $NOW_TOKEN
      env:
        NOW_TOKEN: ${{ secrets.NOW_TOKEN }}
```

If you view the action running, during the `Deploy` step, you should see something like

```bash
Run now build --prod --confirm -t $NOW_TOKEN
Now CLI 17.1.1

> The `name` property in now.json is deprecated (https://zeit.ink/5F)

Reviewer 5: Can we tell them to do something that _isn't_ deprecated?

Author: I can tinker around with that, I got this from someone that works for Zeit. But let me try somethings and update accordingly.

Author: @10xLaCroixDrinker so what they are recommending is project linking. What this does is creates a `.now` folder and links it by id instead of by name. It would require the end-user to commit this `.now` folder which when you run `now` locally it automatically adds it to your `.gitignore`. I was talking to Guillermo Rauch who is the CEO of Zeit and he mentioned that they can work on improving that message. So for the time being to run it in a CI/CD pipeline (CircleCI, GitHub Actions, Travis) you have to run it this way.

The issue with project linking is that we need `NODE_ENV=development` for `npm install` but then `NODE_ENV=production` for the bundling of the module. It looks like in their canary version of `now-cli` they allow you to install all the `devDependencies` even when `NODE_ENV=production`. However, not sure if we want to rely on that.

[👈 Return to Overview](./Recipes.md)

# Deploying Modules

## 📖 Table of Contents

> * [Deploying to Now with GitHub Actions](#deploying-to-now-with-github-actions)

Reviewer 6: is a TOC needed with only 1 item? or this is for consistency ... if its for consistency then ok

Author: I have it for consistency for when we eventually add more ways.

# Deploying to Now with GitHub Actions

[Now.sh](https://zeit.co/home) allows you to deploy statics such as your modules. We will accomplish this be using [GitHub Actions](https://github.com/features/actions).

## Creating your Deploy Action

> This deploy action uses Zeit's `now-cli` which you will need a [token](https://zeit.co/account/tokens) from Now. The below action runs whenever a release is created or a push to master. This will create 

Reviewer 6: For this you will need Zeit's [now-cli](link-to-it-please) and a [token from Now](https://zeit.co/account/tokens).

> This deploy action uses Zeit's `now-cli` which you will need a [token](https://zeit.co/account/tokens) from Now. The below action runs whenever a release is created or a push to master. This will create 

Reviewer 6: >whenever a release is created or a push to master.

>whenever a release is created or a push to master occurs.

```yml
name: Deploy

on:
  release:
    types: [created]
  push:
    branches:
      - master

jobs:
  deploy:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Use Node.js 12.x
      uses: actions/setup-node@v1
      with:
        node-version: '12.x'
    - name: Cache NPM Dependencies
      uses: actions/cache@v1
      with:
        path: ~/.npm
        key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
        restore-keys: |
          ${{ runner.os }}-node-
    - name: Install Dependencies
      run: npm install
      env:
        NODE_ENV: development
    - name: Build One App Page Wrapper
      run: npm run build
      env:
        NODE_ENV: production

>     - name: 'Deploy'

Reviewer 2: Are we covering the generation of the module map in this guide? The deployed assets are unusable if the module map doesn't get updated at the same time because the integrity digests are different every-time you do `npm run build`

Author: Yeah, that was going to be the next section I added. I just wanted to get this up to get feedback on the initial part.

>       run: now build --prod --confirm -t $NOW_TOKEN

Reviewer 2: The static assets will be published from the build folder, is there a way to output the resulting url that nowsh gives you back after deploying? something along the lines of "successfully deployed assets to `https://subdomain.nows.sh`"

Author: So if you watch the action, something like this will appear

```bash
Run now build --prod --confirm -t $NOW_TOKEN
Now CLI 17.1.1
The `name` property in now.json is deprecated (https://zeit.ink/5F)
- Setting up project
Linked to jamesrsingleton/one-app-page-wrapper (created .now and added it to .gitignore)
Inspect: https://zeit.co/jamesrsingleton/one-app-page-wrapper/if0lpn2bd [671ms]
- Building
- Completing
Production: https://one-app-page-wrapper.now.sh [3s]
https://one-app-page-wrapper-if0lpn2bd.now.sh
```

Reviewer 2: So when you create the token you also pick your subdomain?

Author: Token and subdomain are separate. The subdomain is generated by Now off of your repo name. You can go into the settings latter to change that if you have a specific subdomain.

Reviewer 2: that makes sense, in surge you can chose it first and then have a file to redeploy to the same domain

      env:
        NOW_TOKEN: ${{ secrets.NOW_TOKEN }}

> ```

Reviewer 6: little detail on where the asset ends up after the build, then I think this is super clean and to the point
